include/EncoderDecoder/AbstractEncoderDecoder.h

```diff
@@ -0,0 +1,57 @@
+#ifndef ABSTRACT_ENCODER_DECODER_H
+#define ABSTRACT_ENCODER_DECODER_H
+
+#include <string>
+
+template <typename MappingType, typename EncodedType>
+class AbstractEncoderDecoder {
+public:
+	/**
+	 * @brief Encodes a given plaintext string.
+	 *
+	 * @param plaintext String to encode
+	 * @return std::pair<bool, EncodedType> A pair representing whether the
+	 * encoding succeeded and the result of the encoding operation. If the first
+	 * value is false, the value of the second is undefined.
+	 */
+	virtual std::pair<bool, EncodedType> encode(
+		const std::string &plaintext) const = 0;
+
+	/**
+	 * @brief Decodes an encoded input into a string
+	 *
+	 * @param encodedPlaintext An object of type EncodedType representing the
+	 * encoded plaintext
+	 * @return std::pair<bool, std::string> A pair representing whether the
+	 * decoding succeeded and the result of the decoding operation. If the first
+	 * value is false, the value of the second is undefined.
+	 */
+	virtual std::pair<bool, std::string> decode(
+		const EncodedType &encodedPlaintext) const = 0;
+
+	/**
+	 * @brief Extracts the canonical encoding for the given plaintext and sets
+	 * it as the encoding to use for future encode/decode operations
+	 *
+	 * @param plaintext The text to extract the encoding from
+	 * @return true If the encoding was successfuly created
+	 * @return false If failed in creating the encoding
+	 */
+	virtual bool makeEncodingFromText(const std::string &plaintext) = 0;
```

Comment:
Should the `AbstractEncoderDecoder` be aware of the existence of internal encodings and mapping? At first glance it looks like implementation detail and not part of the interface

Reply:
If you read the requirements and the intended usage, then there's a clear distinction between using a canonical encoding (which is constructed from the text), and using an external encoding. This API allows this distinction.
There is the option of constructing the encoding from the text elsewhere and then passing it through `setEncoding`. Do you think it will be wiser? If so where should this code be in?

Reply:
I only meant It seems weird that the “abstract” interface knows about the Huffman encoding mechanism, the existence of this API in the concrete class is fine

Reply:
I think it is an integral part of the API, that applies to any encoding type, so I leave it as is.